tokio-macros/src/entry/parser.rs

```diff
@@ -0,0 +1,252 @@
+                    ));
+                    self.parse_eq()?;
+                    self.parse_literal()?;
+                    None
+                }
+                name => {
+                    self.errors.push(Error::new(ident.span(), format!("unknown option `{}`, expected one of: `flavor`, `worker_threads`, `start_paused`", name)));
+                    None
+                }
+            },
+            tt => {
+                let span = tt.map(|tt| tt.span()).unwrap_or_else(Span::call_site);
+                self.errors.push(Error::new(span, "expected identifier"));
+                None
+            }
+        }
+    }
+
+    /// Parse the next element as a literal value.
+    fn parse_literal(&mut self) -> Option<Literal> {
+        match self.base.bump() {
+            Some(TokenTree::Literal(literal)) => Some(literal),
+            tt => {
+                let span = tt.map(|tt| tt.span()).unwrap_or_else(Span::call_site);
+                self.errors.push(Error::new(span, "expected literal"));
+                None
+            }
+        }
+    }
+
+    /// Parse a token.
+    fn parse_token(&mut self) -> Option<TokenTree> {
+        match self.base.bump() {
+            Some(t) => Some(t),
+            tt => {
+                let span = tt.map(|tt| tt.span()).unwrap_or_else(Span::call_site);
+                self.errors.push(Error::new(span, "expected token"));
+                None
+            }
+        }
+    }
+
+    /// Parse the next element as an `=` punctuation.
+    fn parse_eq(&mut self) -> Option<()> {
+        match self.base.peek_punct()? {
+            p @ Punct { chars: EQ, .. } => {
+                self.base.step(p.len());
+                Some(())
+            }
+            p => {
+                self.errors
+                    .push(Error::new(p.span, "expected assignment `=`"));
+                None
+            }
+        }
+    }
+}
+
+/// A parser for the item annotated with an entry macro.
+pub(crate) struct ItemParser<'a> {
+    base: BaseParser<'a>,
+}
+
+impl<'a> ItemParser<'a> {
+    /// Construct a new parser around the given token stream.
+    pub(crate) fn new(stream: proc_macro::TokenStream, buf: &'a mut Buf) -> Self {
+        Self {
+            base: BaseParser::new(stream, buf),
+        }
+    }
+
+    /// Parse and produce the corresponding token stream.
+    pub(crate) fn parse(mut self) -> ItemOutput {
+        let start = self.base.len();
+        let mut signature = None;
+        let mut block = None;
+
+        let mut has_async = false;
+
+        let mut tail_state = TailState::default();
+
+        while let Some(tt) = self.base.bump() {
+            match tt {
+                TokenTree::Ident(ident) if self.base.buf.display_as_str(&ident) == "async" => {
+                    // NB: intentionally skip over this token.
+                    has_async = true;
+                }
+                TokenTree::Group(g) if matches!(g.delimiter(), Delimiter::Brace) => {
```

Comment:
If the block or async is a macro variable, it will be passed as None-delimited group and parsing will fail.

Additionally, if block is a macro variable and const generics are present in the function signature, the parsing is considered successful, but the span points to the wrong blocks.

Reply:
Interesting!

So for the first part about None-groups.

This currently works, presumably because the attribute is evaluated *after* macro expansion (otherwise `syn` wouldn't work either since it to my understanding doesn't account for None-groups in those locations either):

```rust
macro_rules! test {
    ($async:tt, $block:block) => {
        #[selectme::main]
        $async fn foo() -> bool $block
    }
}

test!(async, {
    true
});

fn main() {
    println!("{}", foo());
}
```

None-groups only seem to be incorporated if some form of expression and context where it's demanded, and it seems like macro expansion rejects such expressions early before allowing them to be passed to an attribute macro:

```rust
macro_rules! test {
    ($async:tt, $($tt:tt)*) => {
        #[selectme::main]
        async fn foo() -> bool $($tt)*
    }
}

test!(async, {
    true
} + 1);
```

```
error: macro expansion ignores token `+` and any following
  --> examples\error.rs:10:3
   |
8  | / test!(async, {
9  | |     true
10 | | } + 1);
   | |___^__- caused by the macro expansion here
   |
   = note: the usage of `test!` is likely invalid in item context

error: could not compile `selectme` due to previous error
```

As for the second, I agree.

Since angle brackets which makes up generics aren't a group they'd have to be checked for and skipped over intentionally since (currently unstable) const expressions could be present 🤔. This might indeed be an instance where it's preferable to use `syn` to parse the signature, but in order to do this conveniently a patch to `syn` which parses the signature while leaving the body as a block intact might be preferable.

Reply:
Just want to clarify that adding support for None-groups should still happen. I was just curious how they were or were not used.

Reply:
Also, another quick and dirty way to find the presumed function body would be to pick the last braced block found. I doubt any tail expressions will be ever added to the language which makes this invalid.

Reply:
Should be resolved with dirty parsing rather than pulling in `syn` again. I added some documentation to the `parse` function explaining the reasoning.